shard/committee/assignment.go

```diff
@@ -35,11 +34,134 @@ type Reader interface {
 
 // StakingCandidatesReader ..
 type StakingCandidatesReader interface {
+	CurrentBlock() *types.Block
 	ReadValidatorInformation(addr common.Address) (*staking.ValidatorWrapper, error)
 	ReadValidatorSnapshot(addr common.Address) (*staking.ValidatorWrapper, error)
 	ValidatorCandidates() []common.Address
 }
 
+// CandidatesForEPoS ..
+type CandidatesForEPoS struct {
+	Orders                             map[common.Address]effective.SlotOrder
+	OpenSlotCountForExternalValidators int
+}
+
+// CompletedEPoSRound ..
+type CompletedEPoSRound struct {
+	MedianStake         numeric.Dec              `json:"epos-median-stake"`
+	MaximumExternalSlot int                      `json:"max-external-slots"`
+	AuctionWinners      []effective.SlotPurchase `json:"epos-slot-winners"`
+	AuctionCandidates   []*CandidateOrder        `json:"epos-slot-candidates"`
+}
+
+// CandidateOrder ..
+type CandidateOrder struct {
+	*effective.SlotOrder
+	Validator common.Address
+}
+
+// MarshalJSON ..
+func (p CandidateOrder) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		*effective.SlotOrder
+		Validator string `json:"validator"`
+	}{
+		p.SlotOrder, common2.MustAddressToBech32(p.Validator),
+	})
+}
+
+// NewEPoSRound runs a fresh computation of EPoS using
+// latest data always
+func NewEPoSRound(stakedReader StakingCandidatesReader) (
+	*CompletedEPoSRound, error,
+) {
+	eligibleCandidate, err := prepareOrders(stakedReader)
+	if err != nil {
+		return nil, err
+	}
+	maxExternalSlots := shard.ExternalSlotsAvailableForEpoch(
+		stakedReader.CurrentBlock().Epoch(),
+	)
+	median, winners := effective.Apply(
+		eligibleCandidate, maxExternalSlots,
+	)
+	auctionCandidates := make([]*CandidateOrder, len(eligibleCandidate))
+
+	i := 0
+	for key := range eligibleCandidate {
+		auctionCandidates[i] = &CandidateOrder{
+			SlotOrder: eligibleCandidate[key],
+			Validator: key,
+		}
+		i++
+	}
+
+	return &CompletedEPoSRound{
+		MedianStake: median,
+		// NOTE + 1 because we use it as index in assigment,
+		// but as count in semantics
+		MaximumExternalSlot: maxExternalSlots + 1,
```

Comment:
I checked maxExternalSlots = (instance.NumNodesPerShard() -
			instance.NumHarmonyOperatedNodesPerShard()) *
			int(instance.NumShards()) 

it's real count. Not index starting from 0. Don't need to + 1.